routes/ballot.js

```diff
@@ -39,8 +51,42 @@ router.get('/', async (req, res) => {
 	res.send(returnJson)
 })
 
-router.post('/', (req, res) => {
+router.post('/', async (req, res) => {
+	// check that the user POSTing is logged in
+	if (!req.user) {
+		res.sendStatus(401) // the user is not logged in
+		return
+	}
+
+	if (!req.election.isOpen) {
+		// the election is not open now. Reject request.
+		res.sendStatus(403) // forbidden
+		return
+	}
+
+	if (req.body) {
+		let ballot = req.body
+		let votePromises = [] // collect all the promises for inserting the data
+		ballot.forEach((vote) => {
+			votePromises.push(models.Vote.create({ // add a promise to the array for each of the votes
+				rank: (req.election.isRankChoice) ? vote.rank : 1, // if the election is rank choice, the user needs to tell us the rank of this vote
+				UserId: req.user.id,
+				ElectionId: req.electionId,
+				PositionId: vote.positionId,
+				CandidateId: vote.candidateId
+			})
+			.error(() => {
+				res.sendStatus(500)
+				return
+			}))
+		})
 
+		await Promise.all(votePromises) // wait for all the votes to be inserted
```

Comment:
This is a key line that will make sure that the user doesn't have to wait forever for their votes to be inserted. I could have used `bulkCreate()`, but this works the same way and is a little easier to write given the structure of the incoming JSON.